core/src/main/java/org/apache/iceberg/avro/AvroSchemaVisitor.java

```diff
@@ -58,7 +58,7 @@
         return visitor.union(schema, options);
 
       case ARRAY:
-        if (schema.getLogicalType() instanceof LogicalMap || AvroSchemaUtil.isKeyValueSchema(schema.getElementType())) {
+        if (schema.getLogicalType() instanceof LogicalMap && AvroSchemaUtil.isKeyValueSchema(schema.getElementType())) {
```

Comment:
I think the solution should be a better check in `isKeyValueSchema`. If the logical type is `LogicalMap`, then it should not need additional requirements for the element type.

Reply:
>>  If the logical type is LogicalMap, then it should not need additional requirements for the element type.

I think that makes sense. In that case. Maybe I should just simplify the if condition to 
```f (schema.getLogicalType())```  and maybe make the `isKeyValueSchema` check as part of  a PreconditionCheck?

Reply:
Yes, that sounds good.

This was originally an or condition because we had manifest files that were written with arrays of structs for maps, but without the map logical type. We still want to support reading files like that (at least in our version) but I think the right way to support it is to use an `AvroSchemaWithType` visitor introduced in #601 and checking whether the Iceberg type is a map.